core/common/src/files/FileSystem.kt

```diff
@@ -0,0 +1,147 @@
+/*
+ * Copyright 2010-2023 JetBrains s.r.o. and Kotlin Programming Language contributors.
+ * Use of this source code is governed by the Apache 2.0 license that can be found in the LICENSE.txt file.
+ */
+
+package kotlinx.io.files
+
+import kotlinx.io.IOException
+import kotlinx.io.Sink
+import kotlinx.io.Source
+
+/**
+ * An interface providing basic operations on a file system.
+ *
+ * **This API is unstable and subject to change.**
+ */
+public sealed interface FileSystem {
+    /**
+     * Path to a directory suitable for creating temporary files.
+     */
+    public val temporaryDirectory: Path
```

Comment:
I wonder if this is a property of the file system or a system in general

Reply:
The current implementation is not good at distinguishing underlying filesystems (it doesn't see the difference between `/tmp` mounted to tmpfs and `/home/usrename` mounted, let's say, to zfs).
So given then `FileSystem.System` instance, it's `temporaryDirectory` is a system-wide temporary directory.

On the other hand, it there will be a FileSystem representing some network smb share or a ssh-fs, then I would expect the different temp directory for it.

Reply:
Is there plan to support file systems projected over arbitrary sources of data? Like an AWS S3 bucket, an Android APK's assets directory, a zip file, or a zip file within another zip file? I don't think a temporary directory necessarily makes sense at that level.

Reply:
Made it static and renamed to `SystemTemporaryDirectory`